include/gsl/span

```diff
@@ -121,139 +122,165 @@ namespace details
         {
-            Expects(index_ != span_->size());
-            return *(span_->data() + index_);
+            return {begin_, end_, current_};
         }
 
-        constexpr pointer operator->() const
+        constexpr reference operator*() const noexcept { return *operator->(); }
+
+        constexpr pointer operator->() const noexcept
         {
-            Expects(index_ != span_->size());
-            return span_->data() + index_;
+            Expects(begin_ && current_ && end_);
+            Expects(current_ < end_);
+            return current_;
         }
-
         constexpr span_iterator& operator++() noexcept
         {
-            Expects(0 <= index_ && index_ != span_->size());
-            ++index_;
+            Expects(begin_ && current_ && end_);
+            Expects(current_ < end_);
+            ++current_;
             return *this;
         }
 
         constexpr span_iterator operator++(int) noexcept
         {
-            auto ret = *this;
-            ++(*this);
+            auto ret{*this};
+            ++*this;
             return ret;
         }
 
         constexpr span_iterator& operator--() noexcept
         {
-            Expects(index_ != 0 && index_ <= span_->size());
-            --index_;
+            Expects(begin_ && current_ && end_);
+            Expects(current_ > begin_);
+            --current_;
             return *this;
         }
 
         constexpr span_iterator operator--(int) noexcept
         {
-            auto ret = *this;
-            --(*this);
+            auto ret{*this};
+            --*this;
             return ret;
         }
 
-        constexpr span_iterator operator+(difference_type n) const noexcept
+        constexpr span_iterator& operator+=(const difference_type n) noexcept
         {
-            auto ret = *this;
+            Expects(begin_ && current_ && end_);
+            if (n > 0) Expects(end_ - current_ >= n);
+            if (n < 0) Expects(end_ - current_ >= -n);
+            current_ += n;
+            return *this;
+        }
+
+        constexpr span_iterator operator+(const difference_type n) const noexcept
+        {
+            auto ret{*this};
             return ret += n;
         }
 
-        friend constexpr span_iterator operator+(difference_type n, span_iterator const& rhs) noexcept
+        friend constexpr span_iterator operator+(const difference_type n,
+                                                 span_iterator const& rhs) noexcept
         {
             return rhs + n;
         }
 
-        constexpr span_iterator& operator+=(difference_type n) noexcept
+        constexpr span_iterator& operator-=(const difference_type n) noexcept
         {
-            Expects((index_ + n) >= 0 && (index_ + n) <= span_->size());
-            index_ += n;
+            Expects(begin_ && end_ && current_);
+            if (n > 0) Expects(end_ - current_ >= n);
+            if (n < 0) Expects(end_ - current_ >= -n);
```

Comment:
Please double-check these Expects. They appear to be wrong (as they are identical to `+=` Expects).

Reply:
I'll verify the Expects tomorrow morning.